### PR TITLE
Track in profiler if thread was attached to JVM and wasn't detached

### PR DESCRIPTION
**Pull request by AGulev**
It detects only cases when thread attached and wasn't detached between frames. 
Not sure how useful it is since we detach thread every from in our input system.

### CODE REVIEW

File: engine/dlib/src/dlib/android.cpp [hunk 1]
```diff
@@ -27,7 +27,6 @@ ThreadAttacher::ThreadAttacher()
 , m_Env(NULL)
 , m_IsAttached(false)
 {
-    JNIEnv* env = 0;
```
Comment (AGulev) [inline engine/dlib/src/dlib/android.cpp:30]: unused local var

File: engine/profiler/src/profiler.cpp [hunk 2]
```diff
@@ -695,6 +695,8 @@ static dmExtension::Result UpdateProfiler(dmExtension::Params* params)
         DM_PROPERTY_SET_U32(rmtp_Memory, dmProfilerExt::GetMemoryUsage() / 1024u);
     }
 
+    dmProfilerExt::UpdatePlatformProfiler();
```
Comment (AGulev) [inline engine/profiler/src/profiler.cpp:698]: platform specific update functionality

File: engine/profiler/src/profiler_android.cpp [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+DM_PROPERTY_EXTERN(rmtp_Profiler);
+DM_PROPERTY_BOOL(rmtp_AttachedToJVM, false, FrameReset, "Thread attached to JVM", &rmtp_Profiler);
+
+extern struct android_app* __attribute__((weak)) g_AndroidApp;
+
+void dmProfilerExt::SampleCpuUsage()
+{
+    // Should be implementd using JNI for Android 8+
+    // see https://github.com/defold/defold/issues/3385
+    dmProfilerExt::SampleProcCpuUsage();
```
Comment (AGulev) [inline engine/profiler/src/profiler_android.cpp:30]: we need a separate android  implementation to implement custom logic for CPU sampling see https://github.com/defold/defold/issues/3385

File: engine/profiler/src/profiler_android.cpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+double dmProfilerExt::GetCpuUsage()
+{
+    return dmProfilerExt::GetProcCpuUsage();
+}
+
+void dmProfilerExt::UpdatePlatformProfiler()
+{
+    // Shows if thread was attached to JVM and wasn't detached
+    JNIEnv* env = 0;
+    DM_PROPERTY_SET_BOOL(rmtp_AttachedToJVM, g_AndroidApp->activity->vm->GetEnv((void **)&env, JNI_VERSION_1_6) == JNI_OK);
```
Comment (AGulev) [inline engine/profiler/src/profiler_android.cpp:47]: I don't attach thread, but if it wasn't detached it stays attached.

File: engine/profiler/src/profiler_linux.cpp [hunk 5]
```diff
@@ -12,118 +12,25 @@
...
-#include <unistd.h>
-
 #include "profiler_private.h"
-
-// Inspired by this post: http://stackoverflow.com/questions/1420426/how-to-calculate-the-cpu-usage-of-a-process-by-pid-in-linux-from-c
-static uint64_t _sample_cpu_last_t = 0;
-static long int _sample_cpu_last_tot = 0;
-static long int _sample_cpu_last_proc = 0;
-static double _sample_cpu_usage = 0.0;
-static bool _sample_cpu_enabled = true;
```
Comment (AGulev) [inline engine/profiler/src/profiler_linux.cpp:25]: Use a separate `proc` utils to be able to use is on linux and android, but android also needs a separate implementation for Android 8+

File: engine/profiler/src/profiler.cpp [hunk 6]
```diff
@@ -28,6 +28,10 @@
 
 #include <algorithm> // std::sort
 
+DM_PROPERTY_GROUP(rmtp_Profiler, "Profiler");
+DM_PROPERTY_U32(rmtp_CpuUsage, 0, FrameReset, "%% Cpu Usage", &rmtp_Profiler);
+DM_PROPERTY_U32(rmtp_Memory, 0, FrameReset, "Memory usage in kb", &rmtp_Profiler);
```
Comment (JCash) [inline engine/profiler/src/profiler.cpp:33]: Why was the code moved?

File: engine/profiler/src/profiler_android.cpp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+#include "profiler_proc_utils.h"
+
+DM_PROPERTY_EXTERN(rmtp_Profiler);
+DM_PROPERTY_BOOL(rmtp_AttachedToJVM, false, FrameReset, "Thread attached to JVM", &rmtp_Profiler);
+
+extern struct android_app* __attribute__((weak)) g_AndroidApp;
+
+void dmProfilerExt::SampleCpuUsage()
+{
+    // Should be implementd using JNI for Android 8+
```
Comment (JCash) [inline engine/profiler/src/profiler_android.cpp:28]: "implemented"

File: engine/profiler/src/profiler_android.cpp (hunk 3 above)
Comment (JCash) [reply engine/profiler/src/profiler_android.cpp:30]: In the future, I think this should be done in a separate PR.

File: engine/profiler/src/profiler_android.cpp (hunk 4 above)
Comment (JCash) [reply engine/profiler/src/profiler_android.cpp:47]: How is this going to be used?

This only checks if the env can be retrieved on _this_ thread (the main thread)?

File: engine/profiler/src/profiler_proc_utils.cpp [hunk 8]
```diff
@@ -0,0 +1,129 @@
+// Copyright 2020-2022 The Defold Foundation
```
Comment (JCash) [inline engine/profiler/src/profiler_proc_utils.cpp:1]: Now I definitely think that these two tasks should be split up in two PR's (android memory usage vs JNI env profile property)

File: engine/profiler/src/profiler_proc_utils.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#ifndef DM_PROFILER_PROC_PRIVATE_H
+#define DM_PROFILER_PROC_PRIVATE_H
+
+#include <script/script.h>
+
+#define SAMPLE_CPU_INTERVAL 0.25
```
Comment (JCash) [inline engine/profiler/src/profiler_proc_utils.h:20]: Is this used anyware outside of the .cpp file?
If not, move it to the .cpp file.

**Review by JCash — CHANGES_REQUESTED** [review]
Split PR into two features.
Also, some more clarification regarding the JNI env profiler property would be useful.

File: engine/profiler/src/profiler_android.cpp (hunk 4 above)
Comment (AGulev) [reply engine/profiler/src/profiler_android.cpp:47]: It's recommended to minimize number of threads which use JNI. In our code and extensions we always use it only from the main thread. 

https://developer.android.com/training/articles/perf-jni

File: engine/profiler/src/profiler_proc_utils.cpp (hunk 8 above)
Comment (AGulev) [reply engine/profiler/src/profiler_proc_utils.cpp:1]: My I wasn't clear in my explanations.

It's just a refactoring. I moved work with `proc` from `profiler_linux` into `profiler_proc_utils` to be able to create `profiler_android` and use proc from both android and linux

File: engine/profiler/src/profiler_android.cpp (hunk 4 above)
Comment (JCash) [reply engine/profiler/src/profiler_android.cpp:47]: "In our code and extensions we always use it only from the main thread."

Is that really true? I guess there is nothing stopping us from attaching a thread on another thread?

File: engine/profiler/src/profiler_proc_utils.cpp (hunk 8 above)
Comment (JCash) [reply engine/profiler/src/profiler_proc_utils.cpp:1]: Sure, but the refactoring isn't related to the android thread profiler variable.

Comment (JCash) [reply engine/profiler/src/profiler_proc_utils.cpp:1]: Most of the code in this PR is about the refactoring, yet the title says something different.

File: engine/profiler/src/profiler_android.cpp (hunk 4 above)
Comment (AGulev) [reply engine/profiler/src/profiler_android.cpp:47]: sure it's not, it's just not recommended. According to the doc we shouldn't use it if it's not necessary

File: engine/profiler/src/profiler_proc_utils.cpp (hunk 8 above)
Comment (AGulev) [reply engine/profiler/src/profiler_proc_utils.cpp:1]: yes, it's just because it's hard to implement andoid related functionality without this refactoring

**Review by JCash — APPROVED** [review]

**Pull request merged**